src/personas/data_manager/mod.rs

```diff
@@ -1195,10 +1154,80 @@ impl DataManager {
         }
     }
 
+    fn handle_fragment_refresh(&mut self,
+                               src: XorName,
+                               fragment: FragmentInfo)
+                               -> Result<(), InternalError> {
+        if self.cache
+               .register_needed_fragment_with_another_holder(fragment.clone(), src) {
+            return Ok(());
+        }
+
+        let holders = match self.fragment_refresh_accumulator
+                  .add(fragment.clone(), src)
+                  .cloned() {
+            Some(holders) => {
+                self.fragment_refresh_accumulator.delete(&fragment);
+                holders
+            }
+            None => return Ok(()),
+        };
+
+        let needed = match fragment {
+            FragmentInfo::ImmutableData(name) => !self.chunk_store.has(&ImmutableDataId(name)),
+            FragmentInfo::MutableDataShell { name, tag, version, .. } => {
+                match self.chunk_store.get(&MutableDataId(name, tag)) {
+                    Err(_) => true,
+                    Ok(data) => data.version() < version,
+                }
+            }
+            FragmentInfo::MutableDataEntry {
+                name,
+                tag,
+                ref key,
+                version,
+                ..
+            } => {
+                if let Ok(data) = self.chunk_store.get(&MutableDataId(name, tag)) {
+                    match data.get(key) {
+                        None => true,
+                        Some(value) if value.entry_version < version => true,
```

Comment:
why not Some(value)  => value.entry_version < version,
then don't need the match arm of Some(_) => false ?